src/main/java/io/strimzi/kafka/bridge/mqtt/core/MqttKafkaMapper.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.kafka.bridge.mqtt.core;
+
+import io.strimzi.kafka.bridge.mqtt.utils.MappingRule;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+
+/**
+ * Responsible for handling all the topic mapping.
+ *
+ * @see MappingRule
+ */
+public class MqttKafkaMapper {
+
+    private static String DEFAULT_KAFKA_TOPIC = "messages_default";
+    private ArrayList<MappingRule> rules;
+    // matches any character except line terminators. Used to replace the placeholders with {something} in the mqtt pattern.
+    private static String PLACEHOLDER_REGEX = "(.*)";
+    // matches any character except a forward slash (/). Used to replace the + in the mqtt pattern.
+    private static String SINGLE_LEVEL_WILDCARD_REGEX = "[^/]+";
+    // matches any character after the string. Used to replace the # in the mqtt pattern.
+    private static String MULTIPLE_LEVEL_WILDCARD_REGEX = ".*";
+
+    /**
+     * Constructor
+     * 
+     * Creates a new instance of MqttKafkaMapper.
+     */
+    public MqttKafkaMapper(ArrayList<MappingRule> rules) {
+        this.rules = rules;
+    }
+
+    /**
+     * Maps an MQTT topic to a Kafka topic. The topic is mapped according to the defined mapping rules.
+     *
+     * @param mqttTopic represents an MQTT topic.
+     * @return a valid Kafka topic.
+     * @see MappingRule
+     */
+    public String map(String mqttTopic) {
+        for (MappingRule rule : this.rules) {
+            //convert the mqtt pattern to a valid regex expression.
+            // the mqtt pattern can contain placeholders like {something}, + and #.
+            // if the mqtt topic contains a +, we replace it with @singleLevelWildcardRegex
+            // if the mqtt topic contains a #, we replace it with @multiLevelWildcardRegex
+            // if the mqtt topic contains a placeholder (pattern \{\w+\}), we replace it with @placeholderRegex
+            String regex = rule.getMqttTopicPattern().replaceAll("\\{\\w+\\}", PLACEHOLDER_REGEX)
+                    .replace("#", MULTIPLE_LEVEL_WILDCARD_REGEX).replace("+", SINGLE_LEVEL_WILDCARD_REGEX);
+            if (mqttTopic.matches(regex)) {
+                String mappedKafkaTopic = rule.getKafkaTopicTemplate();
+                String[] mqttTopicPatternParts = rule.getMqttTopicPattern().split("/");
+                for (String placeholderKey : mqttTopicPatternParts) {
+                    if (placeholderKey.matches("\\{\\w+\\}")) {
+                        int index = Arrays.asList(mqttTopicPatternParts).indexOf(placeholderKey);
+                        String placeholder = mqttTopic.split("/")[index];
+                        mappedKafkaTopic = mappedKafkaTopic.replace(placeholderKey, placeholder);
```

Comment:
So this replacement is not going to work well, you need a more sophisticated logic, at least checking the placeholder names. 
Check this test:

```java
@Test
    public void test() {
        List<MappingRule> rules = new ArrayList<>();
        rules.add(new MappingRule("fleet/{fleet}/vehicle/{vehicle}", "fleet_{flee}"));
        MqttKafkaMapper mapper = new MqttKafkaMapper(rules);
        assertThat(mapper.map("fleet/5/vehicle/10"), is("messages_default"));
    }
```

Why I am expecting to be "message_default"? Because I made a mistake on purpose on the Kafka template ... notice it's `fleet_{flee}` and not `fleet_{fleet}` ... the placeholder name is wrong.
You are just replacing placeholders with the values got from the MQTT pattern but if a placeholder doesn't exist as in this case, it won't be replaced.
In this test, the output Kafka topic is just ... the Kafka template, so it's wrong.

```shell
Expected: is "messages_default"
     but: was "fleet_{flee}"
```

Reply:
Yes, I was thinking about it while implementing. 

In the beginning, I was taking all the placeholders and put them in a hash Map (placeholder name as the key, and the corresponding value from the MQTT topic (input) as the value) and then build the Kafka topic using the Kafka topic template and the hash Map. But the behavior is the same if the key is not found. 

As this should be user defined, I think we can just check the names, if it does not match, we can just return an exception: mapping failed because x was y. 

I can still use the HashMap idea.

Reply:
or maybe, I can check if all the placeholders in the template and in the mqtt pattern are same, if not we can return an error before starting the mapping.

Reply:
I did not use a HashMap.
Just checked if the placeholder in the MQTT pattern is also in the Kafka template. 
I have added tests as well.

Reply:
Just went through the code again and check some issues. I am working on it.